File: 二叉树/后根遍历.js
const postorderTraversal = root => {
  const result = []
  const stack = []
  let T = root
  let flag
  let p = null

  if (T) {
    stack.push(T)
    while (stack.length) {
      while (stack[stack.length - 1]) {
        stack.push(T.lchild)
      }
      stack.pop()
      while (stack.length) {
        T = stack[stack.length - 1]
        if (T.rchild === null || T.rchild === p) {
          stack.pop()
          result.push(T.val)
          p = T
          flag = false
        } else {
          stack.push(T.rchild)
          flag = true
        }
        if (falg) {
          break
        }
      }
    }
  }
}

File: 二叉树/后根遍历.java
// 每个结点访问两次
public int[] postorderTraversal(root) {
  BiTreeNode T = root;
  const result = [];
  if (T != null) {
    LinkStack S = new LinkStack();
    S.push(T);
    // 可能是一棵左子树遍历完, 也有可能是右子树
    Boolean flag;
    // 指向刚刚访问过的结点
    BiTreeNode p = null;
    while (!S.isEmpty()) {
      while (S.peek() != null) {
        S.push(S.peek().lchild);
      }
      S.pop();
      while (!S.isEmpty()) {
        T = S.peek();
        if (T.rchild == null || T.rchild == p) {
          S.pop();
          result.push(T.val);
          p = T;
          flag = false;
        } else {
          S.push(T.rchild);
          flag = true;
        }
        if (flag) {
          break;
        }
      }
    }
  }
  
  return result;
}